rasterio/rio/merge.py

```diff
@@ -34,33 +45,76 @@ def merge(ctx, files, driver):
             files = files[:-1]
 
             with rasterio.open(files[0]) as first:
+                first_res = first.res
                 kwargs = first.meta
-                kwargs['transform'] = kwargs.pop('affine')
-                dest = np.empty((first.count,) + first.shape, 
-                    dtype=first.dtypes[0])
+                kwargs.pop('affine')
 
             if os.path.exists(output):
+                # TODO: prompt user to update existing file (-i option) like:
+                # overwrite b.tif? (y/n [n]) n
+                # not overwritten
                 dst = rasterio.open(output, 'r+')
                 nodataval = dst.nodatavals[0]
+                dest = np.zeros((dst.count,) + dst.shape,
+                        dtype=dst.dtypes[0])
             else:
+                # Create new output file.
+                # Extent from option or extent of all inputs.
+                if not bounds:
+                    # scan input files.
+                    xs = []
+                    ys = []
+                    for f in files:
+                        with rasterio.open(f) as src:
+                            left, bottom, right, top = src.bounds
+                            xs.extend([left, right])
+                            ys.extend([bottom, top])
+                    bounds = min(xs), min(ys), max(xs), max(ys)
+                output_transform = Affine.translation(bounds[0], bounds[3])
+
+                # Resolution/pixel size.
+                if not res:
+                    res = first_res
+                output_transform *= Affine.scale(res[0], -res[1])
+
+                # Dataset shape.
+                output_width = int(math.ceil((bounds[2]-bounds[0])/res[0]))
+                output_height = int(math.ceil((bounds[3]-bounds[1])/res[1]))
+
                 kwargs['driver'] == driver
+                kwargs['transform'] = output_transform
+                kwargs['width'] = output_width
+                kwargs['height'] = output_height
+
                 dst = rasterio.open(output, 'w', **kwargs)
+                dest = np.zeros((first.count,) + (output_height, output_width),
+                        dtype=first.dtypes[0])
                 nodataval = first.nodatavals[0]
 
-            dest.fill(nodataval)
+            if nodataval is not None:
+                dest.fill(nodataval)
 
```

Comment:
@sgillies Consider adding:

```
else:
    nodataval = 0
```

and then remove the ```or 0``` part of the conditionals on lines 111 and 117.  Those seem risky unless handled here, since 0 is only indicating nodata when ```nodataval is None```   (otherwise 0 is possibly a valid pixel value).